### ISSUE TITLE
[iOS 16][video_player] Some videos play only audio with no video

### ISSUE THREAD

**Issue by collinNewCode** [flutter/flutter#111457]
the video_player plugin to play AES-128 encrypted m3u8, the audio is working,but the video is black screen.
It was normal before iOS16.

Unencrypted m3u8 seems to be normal.
Can anyone give me some guidance?

**Comment by danagbemava-nc** [comment]
Hi @collinNewCode,

Please provide your `flutter doctor -v` and a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example).

Also, what version of `video_player` are you using?

Thank you

**Comment by hucool** [comment]
I have the same problem

func copyPixelBuffer always returns null

**Comment by hucool** [comment]
> 

video_player version: 2.4.7

**Comment by collinNewCode** [comment]
`import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

void main() {
  runApp(
    const MaterialApp(
      home: MyVideo(),
    ),
  );
}

class MyVideo extends StatefulWidget {
  const MyVideo({Key? key}) : super(key: key);

  @override
  _MyVideoState createState() => _MyVideoState();
}

class _MyVideoState extends State<MyVideo> {
  VideoPlayerController? _controller;

  @override
  void initState() {
    super.initState();
    init();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("My Video"),
      ),
      body: Center(
        child: _controller!.value.isInitialized
            ? AspectRatio(
                aspectRatio: _controller!.value.aspectRatio,
                child: VideoPlayer(_controller!),
              )
            : Container(
                child: const Text('loading...'),
              ),
      ),
    );
  }

  Future<void> init() async {
    _controller = VideoPlayerController.network(
      'https://service.beta.sanjieke.cn/video/media/11244355/608p.m3u8?user_id=18942194&class_id=33254014&time=1663055202&nonce=868849&token=73daa6bce6e2643a3a9d2f28f3385ce86fa5ec82',
    );

    _controller?.addListener(() {
      setState(() {});
    });

    await _controller?.initialize();

    _controller?.play();
  }
}
`
video_player 版本：2.4.7

**Comment by collinNewCode** [comment]
```
import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

void main() {
  runApp(
    const MaterialApp(
      home: MyVideo(),
    ),
  );
}

class MyVideo extends StatefulWidget {
  const MyVideo({Key? key}) : super(key: key);

  @override
  _MyVideoState createState() => _MyVideoState();
}

class _MyVideoState extends State<MyVideo> {
  VideoPlayerController? _controller;

  @override
  void initState() {
    super.initState();
    init();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("My Video"),
      ),
      body: Center(
        child: _controller!.value.isInitialized
            ? AspectRatio(
                aspectRatio: _controller!.value.aspectRatio,
                child: VideoPlayer(_controller!),
              )
            : Container(
                child: const Text('loading...'),
              ),
      ),
    );
  }

  Future<void> init() async {
    _controller = VideoPlayerController.network(
      'https://service.beta.sanjieke.cn/video/media/11244355/608p.m3u8?user_id=18942194&class_id=33254014&time=1663055202&nonce=868849&token=73daa6bce6e2643a3a9d2f28f3385ce86fa5ec82',
    );

    _controller?.addListener(() {
      setState(() {});
    });

    await _controller?.initialize();

    _controller?.play();
  }
}
```
```
[✓] Flutter (Channel stable, 3.0.5, on macOS 13.0 22A5342f darwin-x64, locale zh-Hans-CN)
    • Flutter version 3.0.5 at /Users/admin/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (9 周前), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[!] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! CocoaPods 1.10.1 out of date (1.11.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] VS Code (version 1.71.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.48.0

[✓] Connected device (2 available)
    • iPhone (mobile) • d2ceb261b299406c50174a51869db35845cf8659 • ios        • iOS 16.0 20A362
    • macOS (desktop) • macos                                    • darwin-x64 • macOS 13.0 22A5342f darwin-x64

[✓] HTTP Host Availability
    • All required HTTP hosts are available
```

video_player version: 2.4.7

@danagbemava-nc

**Comment by danagbemava-nc** [comment]
I can reproduce the issue with the code sample provided above. 
I tested a swiftUI app and everything worked as expected. 


<details>
<summary>videos</summary>

| flutter | swiftUI | 
| -- | -- |
| <video src="https://user-images.githubusercontent.com/88313112/189915121-65439d0f-e8c0-49ea-9d49-1c6eaec4e9cd.mp4" /> | <video src="https://user-images.githubusercontent.com/88313112/189915129-29c61a90-ad03-49d7-b7f3-440849cd29e3.mp4" /> |

</details>


<details>
<summary>swiftUI sample</summary>

```swift
import AVKit
import SwiftUI

struct ContentView: View {
        
    let videoURL = "https://service.beta.sanjieke.cn/video/media/11244355/608p.m3u8?user_id=18942194&class_id=33254014&time=1663055202&nonce=868849&token=73daa6bce6e2643a3a9d2f28f3385ce86fa5ec82"
    @State var showingErrorAlert = false
    @State var errorMessage = ""
    @State var alertTitle = ""
    @State var player : AVPlayer = AVPlayer()
    
    var body: some View {
        VideoPlayer(
            player: player
        ).frame(height: 400)
        .onAppear(perform: playVideo)
        .alert(alertTitle, isPresented: $showingErrorAlert) {
            Button("OK") {}
        } message: {
          Text(errorMessage)
        }
    }
    
    func playVideo() -> Void {
        guard let url = URL(string: videoURL) else {
            alertTitle = "Error"
            errorMessage = "Unable to play video"
            showingErrorAlert = true
            return
        }
        
        player = AVPlayer(url: url)
        
        player.play()
    }
}

struct ContentView_Previews: PreviewProvider {
    static var previews: some View {
        ContentView()
    }
}
```
</details>

<details>
<summary>flutter sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

void main() {
  runApp(
    const MaterialApp(
      home: MyVideo(),
    ),
  );
}

class MyVideo extends StatefulWidget {
  const MyVideo({Key? key}) : super(key: key);

  @override
  State<MyVideo> createState() => _MyVideoState();
}

class _MyVideoState extends State<MyVideo> {
  VideoPlayerController? _controller;

  @override
  void initState() {
    super.initState();
    init();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("My Video"),
      ),
      body: Center(
        child: _controller!.value.isInitialized
            ? AspectRatio(
                aspectRatio: _controller!.value.aspectRatio,
                child: VideoPlayer(_controller!),
              )
            : Container(
                child: const Text('loading...'),
              ),
      ),
    );
  }

  Future<void> init() async {
    _controller = VideoPlayerController.network(
      'https://service.beta.sanjieke.cn/video/media/11244355/608p.m3u8?user_id=18942194&class_id=33254014&time=1663055202&nonce=868849&token=73daa6bce6e2643a3a9d2f28f3385ce86fa5ec82',
    );

    _controller?.addListener(() {
      setState(() {});
    });

    await _controller?.initialize();

    _controller?.play();
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.3.1, on macOS 12.5 21G72 darwin-arm, locale en-GB)
    • Flutter version 3.3.1 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4f9d92fbbd (7 days ago), 2022-09-06 17:54:53 -0700
    • Engine revision 3efdf03e73
    • Dart version 2.18.0
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/212.5712.43.2112.8815526/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A309
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] Android Studio (version 2021.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/212.5712.43.2112.8815526/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.71.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.48.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 4302007A-B132-4B89-887A-C5993FCA3284 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.5 21G72 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 105.0.5195.102

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel master, 3.4.0-19.0.pre.220, on macOS 12.5 21G72 darwin-arm64, locale en-GB)
    • Flutter version 3.4.0-19.0.pre.220 on channel master at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 64b008ab87 (9 hours ago), 2022-09-13 00:33:38 -0400
    • Engine revision 66823c7518
    • Dart version 2.19.0 (build 2.19.0-191.0.dev)
    • DevTools version 2.17.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/212.5712.43.2112.8815526/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A309
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] Android Studio (version 2021.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/212.5712.43.2112.8815526/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.71.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.48.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 4302007A-B132-4B89-887A-C5993FCA3284 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.5 21G72 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 105.0.5195.102

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by collinNewCode** [comment]
> 我可以使用上面提供的代码示例重现该问题。 我测试了一个 swiftUI 应用程序，一切都按预期工作。
> 
> 编辑：我只在 iOS 16 上进行了测试，因为我目前没有 iOS 15 设备
> 
> 视频
> swiftUI 示例
> 颤振样本
> 颤振医生-v
@danagbemava-nc 
Currently this issue only exists in iOS 16.
bro, is there a temporary solution now

**Comment by collinNewCode** [comment]
> 我可以使用上面提供的代码示例重现该问题。 我测试了一个 swiftUI 应用程序，一切都按预期工作。
> 
> 编辑：我只在 iOS 16 上进行了测试，因为我目前没有 iOS 15 设备
> 
> 视频
> swiftUI 示例
> 颤振样本
> 颤振医生-v

@danagbemava-nc
Currently this issue only exists in iOS 16.
bro, is there a temporary solution now

**Comment by danagbemava-nc** [comment]
> bro, is there a temporary solution now

Not that I'm aware of at the moment. Someone from the plugins team will take a look at this soon

**Comment by hellohuanlin** [comment]
I was able to reproduce. The `[_videoOutput hasNewPixelBufferForItemTime:]` always return false in iOS 16.

**Comment by collinNewCode** [comment]
> I was able to reproduce. The `[_videoOutput hasNewPixelBufferForItemTime:]` always return false in iOS 16.

yes，do u have a solution now?

**Comment by hellohuanlin** [comment]
@collinNewCode not yet, but i found [this similar problem](https://stackoverflow.com/questions/42839831/fairplay-streaming-calling-copypixelbufferforitemtime-on-avplayeritemvideooutpu). It was 5 years ago, but could be same problem with just with different format

**Comment by oopuuu** [comment]
I have the same issue, any solutions?

**Comment by JiangJuHong** [comment]
Me too. Video path: [https://yuhong-china.com/20220914/11uas9ID/500kb/hls/index.m3u8](https://yuhong-china.com/20220914/11uas9ID/500kb/hls/index.m3u8)

**Comment by wawo458** [comment]
I have the same problem, is there a solution?

**Comment by Adherentman** [comment]
Is there a solution to this problem? This issue affects a lot of ios16 users。

**Comment by wawo458** [comment]
Has this issue not been taken seriously? affected many users. The problem only occurs when iOS16 plays encrypted video.

**Comment by collinNewCode** [comment]
@danagbemava-nc
bro, is the plugin team working on this issue? 
there is a lot of feedback now

**Comment by hellohuanlin** [comment]
I'm still looking into this. If you find anything that might be helpful, feel free to share as well.

**Comment by wawo458** [comment]
Bro , many people are waiting  , please 😂

**Comment by Jai05-techie** [comment]
Hi team,
 
We are also facing the same issue encrypted videos are not playing on iOS 16 devices properly, only audio is playing.

**Comment by jmagman** [comment]
This is actively being worked on.  Please avoid commenting “me too”, “+1”, or similar on issues. Instead, please add a thumbs-up reaction to the issue to avoid adding noise.  If you’d like to receive updates, click the subscribe button in the right-hand column.

**Comment by hellohuanlin** [comment]
Update: found some leads. Putting up a PR soon.

**Comment by hellohuanlin** [comment]
Update: It worked! 

I am adding some unit tests and will put up a PR soon. But in case you are curious (it's a pretty tricky one), here's some info: 


## Research

There aren't too much info online. It seems to be **by design that only Apple's `AVPlayerLayer` can access those encrypted video stream** on iOS 16. 

For example, [this unresolved issue](https://stackoverflow.com/questions/42839831/fairplay-streaming-calling-copypixelbufferforitemtime-on-avplayeritemvideooutpu) (5 years ago, but could be the same issue under different format): 

> the only way to display your protected video content is by using an AVPlayerLayer

Another hint is that iOS 16 introduced a new API [`AVPlayerLayer::copyDisplayedPixelBuffer`](https://developer.apple.com/documentation/avfoundation/avplayerlayer/4042647-copydisplayedpixelbuffer?language=objc) with a note: 

> It also returns nil when displaying protected content

## Workaround

I almost concluded that there's no way to solve it. But i was wondering "how did Apple solve it in `AVPlayerLayer`"? It must have accessed some private APIs. Something like this: 

```
    if player.videoStream.someMetaData.isProtected {
      player.videoStream._unlockProtectedContent() // a hypothetical private API
    }
```

If I *were* to design AVFoundation, I would probably make it a layered structure, so that the "unlocked" pixel buffers can be piped downstream. This mean that "unlocking" probably happens towards the upstream, hence all outputs (including ours!) can be granted the access. Something like this:  

```mermaid
graph TD;
  A[upstream input]-->B[...];
  B-->C[_unlock, hypothetical private API used by AVPlayerLayer];
  C-->D[...];
  D-->E[...]
  E-->F[output 1];
  E-->G[output 2];
  E-->H[our output];
```

To verify this hypothesis, I created a dummy `AVPlayerLayer` (without adding it to screen), hoping that the hypothetical unlocking happens in the constructor. The result is that **only first video frame is displayed, and all other frames are still not accessible**. 

This tells 3 things: 
1. The above layered structure is likely what Apple is doing 
2. The hypothetical unlocking did happen in the constructor, but only for the first frame. 
3. The "unlocking" process is likely for each specific frame, and not the whole stream

Now based on all the above info, this is what I *guess* Apple's implementation of AVPlayerLayer is: 

```
class AVPlayerLayer { 
  // `init` only unlocks the first frame 
  init(player: AVPlayer) {
    player.videoStream.firstFrame()._unlockFrame() // hypothetical private API
  }

  // each subsequent frame is unlocked separately
  func renderNextFrame() { 
    player.videoStream.frameAtCurrentTime()._unlockFrame() // hypothetical private API

    // now these pixel buffers become accessible
    let pixelBuffer = ... 
    paintOnCanvas(pixelBuffer)
  }
}
```

So in conclusion, **all we need to do is to have a dummy AVPlayerLayer to concurrently play the video**, which should invoke this hypothetical `renderNextFrame()` function, and eventually unlock every single frame for us. 

And it worked :)

![ios 16 video](https://user-images.githubusercontent.com/41930132/190715284-a47ca7c8-164f-4728-a13c-60c6f94a4822.gif)


## Future

It is pretty obvious that Apple's intention is to only allow `AVPlayerLayer` to access those pixel buffers from iOS 16 onwards. 

To follow this spirit, in the future we should probably rely on platform views with `AVPlayerLayer` attached. This would be a huge change from the current implementation though.

**Comment by jmagman** [comment]
> all we need to do is to have an invisible AVPlayerLayer to concurrently play the video

Are there memory or performance implications of doing this?

**Comment by hellohuanlin** [comment]
> Are there memory or performance implications of doing this?

@jmagman my guess is very minimal because it's not rendered on screen. But let me try and compare.

**Comment by hellohuanlin** [comment]
I loop the video for around 20 minutes, and confirmed minimal performance impact: 

Memory without the workaround: 

![wihtout workaround memory](https://user-images.githubusercontent.com/41930132/190804677-845f9e81-1268-4810-be19-2bf878e12441.png)

Memory with the workaround: 

![memory with workaround](https://user-images.githubusercontent.com/41930132/190805012-7583b548-97b1-453b-bbfa-51536097e191.png)


CPU without the workaround: 

![without workaround cpu](https://user-images.githubusercontent.com/41930132/190804284-69e873a3-b372-4b2a-a1cf-97f444384d8c.png)

CPU with the workaround: 

![with workaround cpu](https://user-images.githubusercontent.com/41930132/190804425-ec80538a-3427-4c69-ba3b-3e8e379f3487.png)

**Comment by Adherentman** [comment]
Approximately when the new version will be released？

**Comment by hellohuanlin** [comment]
@Adherentman it should already be released when the PR is landed. Can you update and see if the issue is fixed?

**Comment by Adherentman** [comment]
> @Adherentman it should already be released when the PR is landed. Can you update and see if the issue is fixed?

What do I need to do to test?
In pub.dev I checked video_player plugin version 2.4.7.

The version of video_player_avfoundation in the video_player plugin of the main branch in the github repository is 2.2.17

**Comment by hellohuanlin** [comment]
@Adherentman Hmmm, I saw that video_player_avfoundation in pub.dev is still 2.3.5 ([link](https://pub.dev/packages/video_player_avfoundation/versions)). For some reason my update ([2.3.6](https://github.com/flutter/plugins/blob/main/packages/video_player/video_player_avfoundation/CHANGELOG.md)) is not published immediately. I thought the package is always auto released? Maybe the release process is broken? Let me take a look.

**Comment by jmagman** [comment]
2.3.6 is now available.  It was delayed due to https://github.com/flutter/plugins/actions/runs/3109769619 which failed on `ios-platform_tests CHANNEL:master PACKAGE_SHARDING:--shardIndex 3 --shardCount 4: completed (failure)`, so the flaky integration test that you already turned off.

**Comment by hellohuanlin** [comment]
@jmagman Gotcha. The website still shows 2.3.5 though: https://pub.dev/packages/video_player_avfoundation/versions

**Comment by hellohuanlin** [comment]
FUI there are some issue releasing it. It's being discussed in the discord channel. The sheriff will release it tomorrow.

**Comment by stuartmorgan** [comment]
`video_player_avfoundation` 2.3.6 is published now; `flutter pub upgrade` will get the new version.

**Comment by Adherentman** [comment]
It's work. Thanks!
